src/pops.cpp

```diff
@@ -127,37 +127,20 @@ List pops_model(int random_seed,
                 int latency_period = 0
 )
 {
-  Config config;
+  // TODO no establishment_probability, establishment_stochasticity, or establishment_probability
+  // steps and use_treatments set later (set to 0 and false for now)
+  Config config = {random_seed, num_rows, num_cols, ew_res, ns_res, 0, true, true, 0.5, 
+  use_lethal_temperature, lethal_temperature, weather, reproductive_rate, model_type_, latency_period,
+  natural_kernel_type, natural_distance_scale, natural_dir, natural_kappa, use_anthropogenic_kernel,
+  percent_natural_dispersal, anthropogenic_kernel_type, anthropogenic_distance_scale, anthropogenic_dir,
+  anthropogenic_kappa, false, mortality_on, mortality_rate, mortality_time_lag};
```

Comment:
For better or worse, this won't be possible in the future with Config. Unlike Python, we can make clear key=value assignments in this context (at least with current and foreseeable C++) which makes it error prone (we still need to use these long lists of parameters elsewhere, but well...). Additionally, we will have to use more setters and getters in the Config class to keep some of the scheduling in consistent state. My original idea was that Config will be just a simple struct, but it will have to be a more complex class. However, the scheduling and configuration are interdependent, so my current solution for that is constructing Config in phases. Fixing that in the future may simplify Config class, but it won't fix problems with a long parameter list.

Reply:
Just to clarify, for now I need to initialize each of the public variables separately (like I had it before) and then change these to calling setters when this functionality is expanded? Would it be possible to add a non-generic constructor to Config to avoid setting so many variables directly in pops.cpp?

Reply:
The answer to your first question is yes. I don't understand the second one. Do you want to prepare some pseudo-code of what you are envisioning? There should be a constructor with minimalistic defaults, so that the code does not crash when these are used, but all variables need to be set anyway because user needs to be able to set them, so I guess that's not what you mean.

Reply:
I was thinking that instead of having a generic constructor for the config class [i.e Config() {...}] it could have a constructor with parameters that are currently the public variables in the config class [i.e Config(int random_seed, int rows, int cols, double ew_res, double ns_res, bool generate_stochasticity, .... ){...}] that way in the pops.cpp class instead of having ~20 lines of setting variables (i.e config.natural_kernel_type = natural_kernel_type; and so on) it would just call a constructor [Config config(seed, row, col, ew_res, ...)]. But, I was looking over the r.pops.spread main.cpp and I see that that approach does not work as well there.

Reply:
Right, it does not, but that is only part of the issue. The other part is that it is hard to get a long list of parameters right in C++ since you can't use their names and everything relies on order and type, but if you have a lot of variables with the same type, then it is really just the order. It is also less readable overall. Reducing number of lines would be nice, but not in this way. Compare readability and potential for errors in these four cases:

```
config1 = Config(temperature, year, month, day, num_steps);  // let's say this is correct and fits the def

config2 = Config();  // we can see right here that this is correct
config2.temperature = temperature;
config2.year = year;
config2.month = month;
config2.day = day;
config2.num_steps = num_steps;

config3 = Config(temperature, num_steps, year, month, day);  // oops, wrong order 

config4 = Config();
config4.temperature = num_steps;
config4.year = year;
config4.month = month;
config4.day = day;
config4.num_steps = temperature;  // pretty clear I made a copy-paste mistake here
```

I originally aimed at API with these public member assignments, but that won't work for things around scheduling and time. Setters are next closest solution, so we will move to those, although they have shortcomings such as longer code (not only lines, but also characters) and need for different states and checks in the object not needed if there is only a constructor (and thus one stage (only) initialization).

There are other ways in C++ how to solve this, but they are much less common. In Python and future C++20, there is a way of using key=value syntax in the constructor, but that's not available to us yet, I'm not sure if it would be actually useful as it would not address all our issues, and it is often written line-by-line anyway (e.g. with Python's Black formatting), so resulting in the same number of lines on the caller side.

This basically goes into review of https://github.com/ncsu-landscape-dynamics/PoPS/pull/70, so thanks for starting that already.

Reply:
Thank you for the explanation, I understand now.